jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt

```diff
@@ -0,0 +1,132 @@
+                repeat(propertySize) {
+                    val key = dataInput.readUTF()
+                    val value = dataInput.readUTF()
+                    mutableMap[key] = value
+                }
+                resources.add(IndexedResource(mutableMap))
+            }
+            return resources
+        }
+    }
+
+    override fun getName(): ID<String, MutableList<IndexedResource>> = NAME
+
+    override fun getIndexer(): DataIndexer<String, MutableList<IndexedResource>, FileContent> = DataIndexer { fileContent ->
+        val indexedResources = mutableMapOf<String, MutableList<IndexedResource>>()
+
+        fileContent.psiFile.acceptNode(object : PsiElementVisitor() {
+            override fun visitElement(element: PsiElement?) {
+                super.visitElement(element)
+                element?.run {
+                    val parent = element.parent as? YAMLKeyValue ?: return
+                    if (parent.value != this) return
+
+                    val resource = YamlCloudFormationTemplate.convertPsiToResource(parent) ?: return
+                    val resourceType = resource.type() ?: return
+                    IndexedResource.fromResource(resource)?.let {
+                        indexedResources.computeIfAbsent(resourceType) { mutableListOf() }.add(it)
+                    }
+                }
+            }
+        })
+
+        indexedResources
+    }
+
+    override fun getKeyDescriptor(): KeyDescriptor<String> = EnumeratorStringDescriptor.INSTANCE
+
+    override fun getVersion(): Int = 1
+
+    override fun getInputFilter(): FileBasedIndex.InputFilter = fileFilter
+
+    override fun dependsOnFileContent(): Boolean = true
+
+    companion object {
+        private val NAME: ID<String, MutableList<IndexedResource>> = ID.create("SamTemplateIndex")
+
+        private fun PsiElement.acceptNode(visitor: PsiElementVisitor) {
+            accept(visitor)
+
+            if (children.isNotEmpty()) {
+                children.forEach { it.acceptNode(visitor) }
+            }
+        }
+
+        private fun listAllResources(project: Project, filter: (String) -> Boolean): Collection<IndexedResource> {
+            val index = FileBasedIndex.getInstance()
+            return index.getAllKeys(NAME, project)
+                    .asSequence()
+                    .filter(filter)
+                    .mapNotNull { index.getValues(NAME, it, GlobalSearchScope.projectScope(project)) }
+                    .filter { it.isNotEmpty() }
+                    .flatten()
+                    .flatten()
+                    .toList()
+        }
+
+        fun listResourcesByType(project: Project, type: String): Collection<IndexedResource> = listAllResources(project) { it == type }
+
+        fun listFunctions(project: Project): Collection<IndexedFunction> =
+                listAllResources(project) { it == SERVERLESS_FUNCTION_TYPE || it == LAMBDA_FUNCTION_TYPE }
+                        .asSequence()
```

Comment:
Why do `asSequence` if we're just going to convert it immediately to a list?

Reply:
Ops, removed.